src/ssl.c

```diff
@@ -5173,72 +5179,98 @@ int wolfSSL_Init(void)
         /* Initialize crypto for use with TLS connection */
         if (wolfCrypt_Init() != 0) {
             WOLFSSL_MSG("Bad wolfCrypt Init");
-            return WC_INIT_E;
+            ret = WC_INIT_E;
         }
 
 #ifdef HAVE_GLOBAL_RNG
-        if (wc_InitMutex(&globalRNGMutex) != 0) {
+        if ((ret == WOLFSSL_SUCCESS) && (wc_InitMutex(&globalRNGMutex) != 0)) {
             WOLFSSL_MSG("Bad Init Mutex rng");
-            return BAD_MUTEX_E;
+            ret = BAD_MUTEX_E;
+        }
+        else {
+            globalRNGMutex_valid = 1;
         }
 #endif
 
 #ifdef OPENSSL_EXTRA
     #ifdef HAVE_ATEXIT
         /* OpenSSL registers cleanup using atexit */
-        if (atexit(AtExitCleanup) != 0) {
+        if ((ret == WOLFSSL_SUCCESS) && (atexit(AtExitCleanup) != 0)) {
```

Comment:
Consider moving the `atexit` to only success case... Since the cleanup is always handled now at the bottom.

Reply:
If `atexit()` fails, we still want to call cleanup below, right?